core/src/main/scala/io/projectglow/gff/GffDataSource.scala

```diff
@@ -0,0 +1,245 @@
+
+
+class GffDataSource extends RelationProvider with SchemaRelationProvider with DataSourceRegister {
+
+  override def shortName(): String = "gff"
+
+  /**
+   * Creates relation
+   *
+   * @param sqlContext spark sql context
+   * @param parameters parameters for job
+   * @return Base relation
+   */
+  override def createRelation(sqlContext: SQLContext, parameters: Map[String, String]): BaseRelation = {
+
+    createRelation(sqlContext, parameters, null)
+  }
+
+  /**
+   * Creates relation with user schema
+   *
+   * @param sqlContext spark sql context
+   * @param parameters parameters for job
+   * @param schema     user defined schema
+   * @return Base relation
+   */
+  override def createRelation(sqlContext: SQLContext, parameters: Map[String, String], schema: StructType): BaseRelation = {
+    val path = parameters.get("path")
+    path match {
+      case Some(p) =>
+        val requiredSchema = if (schema == null) {
+          inferSchema (sqlContext, p)
+        } else {
+          schema
+        }
+
+        new GffResourceRelation(sqlContext, p, requiredSchema)
+      case _ => throw new IllegalArgumentException("Path is required")
+    }
+  }
+
+}
+
+
+class GffResourceRelation(
+  val sqlContext: SQLContext,
+  path: String,
+  requiredSchema: StructType
+) extends BaseRelation with TableScan  { // TODO: investigate use of PrunedFilteredScan
+
+  private val spark = sqlContext.sparkSession
+
+
+  override def schema: StructType = {
+    requiredSchema
+  }
+
+
+  override def buildScan(): RDD[Row] = {
+    val csvDf = spark
+      .read
+      .option("sep", COLUMN_DELIMITER)
+      .schema(GffDataSource.gffCsvSchema)
+      .csv(path)
+
+    val dfWithMappedAttributes = convertAttributesToMap(
+      filterCommentsAndFasta(
+        csvDf
+      )
+    )
+
+    requiredSchema.foldLeft(dfWithMappedAttributes)((df, f) =>
+      f match {
+        case f if structFieldsEqualExceptNullability(f, startField) =>
+          df.withColumn(
+            f.name,
+            col(f.name).cast(f.dataType) - 1
+          )
+        case f if structFieldsEqualExceptNullability(f, endField) =>
+          df.withColumn(
+            f.name,
+            col(f.name).cast(f.dataType)
+          )
+        case f if structFieldsEqualExceptNullability(f, scoreField) |
+          structFieldsEqualExceptNullability(f, strandField) |
+          structFieldsEqualExceptNullability(f, phaseField) =>
+          df.withColumn(
+            f.name,
+            when(
+              col(f.name) =!= lit(NULL_IDENTIFIER),
+              col(f.name).cast(f.dataType)
+            ).otherwise(lit(null))
+          )
+        case f => // TODO: To be completed (also rectify column names)
+          df.withColumn(
+            f.name,
+            element_at(col(attributesMapColumnName), f.name)
+          )
+      }
+    )
+      .drop(attributesField.name, attributesMapColumnName) // TODO: Complete the option of keeping attributes field
+      .rdd
```

Comment:
For an easy performance gain, we can override [needConversion](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala#L211-L223) and return `queryExecution.toRdd`.

Reply:
Great. Done